app/src/main/java/org/thoughtcrime/securesms/ringrtc/Camera.java

```diff
@@ -0,0 +1,329 @@
+  @NonNull  private       CameraState.Direction     activeDirection;
+            private       boolean                   enabled;
+            private       boolean                   isInitialized;
+            private       int                       orientation;
+
+  public Camera(@NonNull Context context,
+                @NonNull CameraEventListener cameraEventListener,
+                @NonNull EglBaseWrapper eglBase,
+                @NonNull CameraState.Direction desiredCameraDirection)
+  {
+    this.context                = context;
+    this.cameraEventListener    = cameraEventListener;
+    this.eglBase                = eglBase;
+    CameraEnumerator enumerator = getCameraEnumerator(context);
+    cameraCount                 = enumerator.getDeviceNames().length;
+
+    CameraState.Direction firstChoice = desiredCameraDirection.isUsable() ? desiredCameraDirection : FRONT;
+
+    CameraVideoCapturer capturerCandidate = createVideoCapturer(enumerator, firstChoice);
+    if (capturerCandidate != null) {
+      activeDirection = firstChoice;
+    } else {
+      CameraState.Direction secondChoice = firstChoice.switchDirection();
+      capturerCandidate = createVideoCapturer(enumerator, secondChoice);
+      if (capturerCandidate != null) {
+        activeDirection = secondChoice;
+      } else {
+        activeDirection = NONE;
+      }
+    }
+    capturer = capturerCandidate;
+  }
+
+  @Override
+  public void initCapturer(@NonNull CapturerObserver observer) {
+    if (capturer != null) {
+      eglBase.performWithValidEglBase(base -> {
+        capturer.initialize(SurfaceTextureHelper.create("WebRTC-SurfaceTextureHelper", base.getEglBaseContext()),
+                            context,
+                            new CameraCapturerWrapper(observer));
+        capturer.setOrientation(orientation);
+        isInitialized = true;
+      });
+    }
+  }
+
+  @Override
+  public boolean hasCapturer() {
+    return capturer != null;
+  }
+
+  @Override
+  public void flip() {
+    if (capturer == null || cameraCount < 2) {
+      throw new AssertionError("Tried to flip the camera, but we only have " + cameraCount + " of them.");
+    }
+    activeDirection = PENDING;
+    capturer.switchCamera(this);
+  }
+
+  @Override
+  public void setOrientation(@Nullable Integer orientation) {
+    this.orientation = orientation;
+
+    if (isInitialized && capturer != null) {
+      capturer.setOrientation(orientation);
+    }
+  }
+
+  @Override
+  public void setEnabled(boolean enabled) {
+    Log.i(TAG, "setEnabled(): " + enabled);
+
+    this.enabled = enabled;
+
+    if (capturer == null) {
+      return;
+    }
+
+    try {
+      if (enabled) {
+        Log.i(TAG, "setEnabled(): starting capture");
+        capturer.startCapture(1280, 720, 30);
+      } else {
+        Log.i(TAG, "setEnabled(): stopping capture");
+        capturer.stopCapture();
+      }
+    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)